### PR TITLE
Add basic pull diagnostics support

### LINKED ISSUE #4276: Support pull diagnostics (`textDocument/diagnostic`)
[Pull diagnostics](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_pullDiagnostics) are a relatively new feature in the LSP specification where the client requests the server for diagnostics, instead of the server notifying the client. Known servers that use this feature include [ruby-lsp](https://github.com/Shopify/ruby-lsp) and the new [Roslyn language server](https://github.com/dotnet/roslyn/tree/main/src/Features/LanguageServer) for .NET.

New capabilities:

- `textDocument.diagnostic` (client)
- `diagnosticProvider` (server)

New methods:

- [`textDocument/diagnostic`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_diagnostic) - Computes diagnostics for a single document, and possibly related documents
- [`workspace/diagnostic`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_diagnostic) - Computes workspace-wide diagnostics which were previously published from server to client
- [`workspace/diagnostic/refresh`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnostic_refresh) - Refreshes workspace-wide diagnostics, used when a project-related change is detected

Related:

https://github.com/joaotavora/eglot/discussions/1290
https://github.com/neovim/neovim/issues/22838
https://github.com/neovim/neovim/pull/24128

### PR DESCRIPTION
**Pull request by Ruin0x11**
Only basic support for now; no extra features like related documents. It's just enough to get diagnostics showing with the Roslyn server.

Closes #4276.

### CODE REVIEW

**Comment by jcs090218** [comment]
Can you resolve conflicts and rebase this? Thank you!

**Comment by catssay** [comment]
The newest version of eslint plugin also uses the feature. My project using latest eslint version 9.5.0, it doesn't work with eslint plugin version 2.2.2 which is lastest version lsp-eslint provided. So I downloaded the latest eslint plugin from vscode market and replaced it, but it not work. Then I applied the patch, it worked.

**Comment by Ruin0x11** [comment]
@jcs090218 Could you please take a look at this when you have some time? Thank you.

**Comment by sin-ack** [comment]
I just tested this patch and it doesn't seem to work with Roslyn language server version 4.11.0-3.24320.2. Weirdly enough, I don't see a `textDocument/diagnostic` message sent, neither by opening a file or changing it.

**Comment by jcs090218** [comment]
> @jcs090218 Could you please take a look at this when you have some time? Thank you.

I'm sorry for not getting back to you sooner! I was a little busy recently.

Others have helped test this patch, but it doesn't work. 🤔 Can you confirm this?

**Comment by jcs090218** [comment]
There are compile errors:

```
In lsp-diagnostics--apply-pull-diagnostics:
lsp-mode.el:2340:2: Warning: docstring has wrong usage of unescaped single quotes (use \= or different quoting)

In lsp--after-change:
lsp-mode.el:4929:9: Warning: ‘mapcar’ called for effect; use ‘mapc’ or ‘dolist’ instead
```

File: lsp-mode.el [hunk 1]
```diff
@@ -2299,6 +2300,15 @@ Common usecase are:
...
   (gethash (lsp--fix-path-casing path) lsp-diagnostic-stats))
 
+(defun lsp-diagnostics--request-pull-diagnostics (workspace)
+  (when (lsp-feature? "textDocument/diagnostic")
+    (let ((path (lsp--fix-path-casing (buffer-file-name))))
+      (lsp-request-async "textDocument/diagnostic"
+                         (list :textDocument (lsp--text-document-identifier))
+                         (-lambda ((&DocumentDiagnosticReport :kind :items?))
+                           (lsp-diagnostics--apply-pull-diagnostics workspace path kind items?))
+                         :mode 'tick))))
```
Comment (jcs090218) [inline lsp-mode.el:2310]: Can you add docstring for this and other functions as well? Thanks!

**Comment by Ruin0x11** [comment]
For me at least, I do get diagnostics with csharp-roslyn server:

ss

I'm using version 4.9.0-3.23604.10 of that server on Windows.

**Review by jcs090218 — APPROVED** [review]

**Comment by jcs090218** [comment]
Thank you!

**Pull request merged**